types/occ/scheduler.go

```diff
@@ -18,3 +23,147 @@ func NewEstimateAbort(dependentTxIdx int) Abort {
 		Err:            ErrReadEstimate,
 	}
 }
+
+type TaskStatus string
+
+const (
+	TaskStatusPending   TaskStatus = "pending"
+	TaskStatusExecuted  TaskStatus = "executed"
+	TaskStatusAborted   TaskStatus = "aborted"
+	TaskStatusValidated TaskStatus = "validated"
+)
+
+type Task struct {
+	Status      TaskStatus
+	Index       int
+	Incarnation int
+	Request     types.RequestDeliverTx
+	Response    *types.ResponseDeliverTx
+}
+
+// TODO: (TBD) this might not be necessary to externalize, unless others
+func NewTask(request types.RequestDeliverTx, index int) *Task {
+	return &Task{
+		Request: request,
+		Index:   index,
+		Status:  TaskStatusPending,
+	}
+}
+
+// Scheduler processes tasks concurrently
+type Scheduler interface {
+	ProcessAll(ctx sdk.Context, tasks []*Task) error
+}
+
+type scheduler struct {
+	deliverTx func(ctx sdk.Context, req types.RequestDeliverTx) (res types.ResponseDeliverTx)
+	workers   int
+}
+
+// NewScheduler creates a new scheduler
+func NewScheduler(workers int, deliverTxFunc func(ctx sdk.Context, req types.RequestDeliverTx) (res types.ResponseDeliverTx)) Scheduler {
+	return &scheduler{
+		workers:   workers,
+		deliverTx: deliverTxFunc,
+	}
+}
+
+func (s *scheduler) ProcessAll(ctx sdk.Context, tasks []*Task) error {
+	toExecute := tasks
+	for len(toExecute) > 0 {
+
+		// execute sets statuses of tasks to either executed or aborted
+		err := s.executeAll(ctx, toExecute)
+		if err != nil {
+			return err
+		}
+
+		// validate returns any that should be re-executed
+		// note this processes ALL tasks, not just those recently executed
+		toExecute, err = s.validateAll(ctx, tasks)
+		if err != nil {
+			return err
+		}
+		for _, t := range toExecute {
+			t.Incarnation++
+			t.Status = TaskStatusPending
+			//TODO: reset anything that needs resetting
+		}
+	}
+	return nil
+}
+
+// TODO: validate each task
+// TODO: return list of tasks that are invalid
+func (s *scheduler) validateAll(ctx sdk.Context, tasks []*Task) ([]*Task, error) {
+	var res []*Task
+	for _, t := range tasks {
+		// any aborted tx is known to be suspect here
+		if t.Status == TaskStatusAborted {
+			res = append(res, t)
+		} else {
+			//TODO: validate the task and add it if invalid
+			//TODO: create and handle abort for validation
+			t.Status = TaskStatusValidated
+		}
+	}
+	return res, nil
+}
+
+// ExecuteAll executes all tasks concurrently
+// Tasks are updated with their status
+// TODO: retries on aborted tasks
+// TODO: error scenarios
+func (s *scheduler) executeAll(ctx sdk.Context, tasks []*Task) error {
+	ch := make(chan *Task, len(tasks))
+	grp, gCtx := errgroup.WithContext(ctx.Context())
```

Comment:
shouldn't we create the context `WithCancel` so that we can kill off the workers from the scheduler? we check the context.Done case below, but we would still need to indicate `Done` for the context via Cancel right?

Reply:
With errgroups, one can just return an error like context timeout error, and all ctx.Done()s will cancel.  If something outside a task needs to cancel, then yes we can wrap errgroup with a WithCancel context, but then it's still going to be the ctx.Done() that exits those routines.